### PR TITLE
[runtime] Fix jit-info.c memory consistency

### PR DESCRIPTION
**Pull request by alexanderkyte**
In the process of getting the tests added in https://github.com/mono/mono/pull/12125 green, I ran into a lot of data races and use-after-frees in jit-info.c . 

This PR has extracted them, so they can be considered separately.

### CODE REVIEW

**Comment by alexanderkyte** [comment]
Using jemalloc with `--enable-fill ` is helpful for forcing these use-after-frees to crash quickly. It fills memory with a single value when it's freed.

**Comment by alexanderkyte** [comment]
```* thread #1, name = 'tid_307', queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x7fa62ea1ed28)
    frame #0: 0x00000001037c9f47 mono-sgen`mono_get_hazardous_pointer(pp=0x00007fa62ea1ed28, hp=0x0000000103b3a600, hazard_index=1) at hazard-pointer.c:208
   205 	
   206 		for (;;) {
   207 			/* Get the pointer */
-> 208 			p = *pp;
   209 			/* If we don't have hazard pointers just return the
   210 			   pointer. */
   211 			if (!hp)
Target 0: (mono-sgen) stopped.

Executable module set to "/Users/akyte/mono_mkbundle/mono/mini/mono-sgen".
Architecture set to: x86_64h-apple-macosx.
(lldb) up
frame #1: 0x00000001035ce04e mono-sgen`jit_info_table_chunk_index(chunk=0x00007fa4ed60abd0, hp=0x0000000103b3a600, addr=0x00007fff73bffb66) at jit-info.c:200
   197 	
   198 		while (left < right) {
   199 			int pos = (left + right) / 2;
-> 200 			MonoJitInfo *ji = (MonoJitInfo *)mono_get_hazardous_pointer((gpointer volatile*)&chunk->data [pos], hp, JIT_INFO_HAZARD_INDEX);
   201 			gint8 *code_end = (gint8*)ji->code_start + ji->code_size;
   202 	
   203 			if (addr < code_end)
(lldb) up
frame #2: 0x00000001035cccba mono-sgen`jit_info_table_find(table=0x00007fa4ed5027e0, hp=0x0000000103b3a600, addr=0x00007fff73bffb66) at jit-info.c:222
   219 		chunk_pos = jit_info_table_index (table, (gint8*)addr);
   220 		g_assert (chunk_pos < table->num_chunks);
   221 	
-> 222 		pos = jit_info_table_chunk_index (table->chunks [chunk_pos], hp, (gint8*)addr);
   223 	
   224 		/* We now have a position that's very close to that of the
   225 		   first element whose end address is higher than the one
(lldb) up
warning: could not execute support code to read Objective-C class data in the process. This may reduce the quality of type information available.
frame #3: 0x00000001035ccaa6 mono-sgen`mono_jit_info_table_find_internal(domain=0x00007fa4ed413100, addr=0x00007fff73bffb66, try_aot=1, allow_trampolines=1) at jit-info.c:293
   290 		   worry about the writer freeing the table. */
   291 		table = (MonoJitInfoTable *)mono_get_hazardous_pointer ((gpointer volatile*)&domain->jit_info_table, hp, JIT_INFO_TABLE_HAZARD_INDEX);
   292 	
-> 293 		ji = jit_info_table_find (table, hp, (gint8*)addr);
   294 		if (hp)
   295 			mono_hazard_pointer_clear (hp, JIT_INFO_TABLE_HAZARD_INDEX);
   296 		if (ji && ji->is_trampoline && !allow_trampolines)
```

**Comment by vargaz** [comment]
The changes look harmless, just not sure what is the exact race they are fixing, and why do those races occur.

**Comment by alexanderkyte** [comment]
@monojenkins build failed

**Comment by alexanderkyte** [comment]
I observed the chunks being disposed prematurely without these changes (filled with the magic value that jemalloc uses on free), as well as entire tables being disposed prematurely sometimes. 

What happened seemed to be lack of coordination around these refcount fields, leading to missed increments / premature freeing. I have not validated this hypothesis beyond applying the synchronization fixes above, and watching the crash reproduction disappear. 

On master, it seems possible to get a lot of stress test runs now without seeing this crash. I've made some of the early steps not use malloc that used to, which makes it not allocate over the freed table slots. 

If you run the malloc crash reproduction in the stress test (`free(-1)`, which asserts on entry), rather than the assert or managed fault ones, you see this more often. Using jemalloc with `enable-fill` also helps see it faster.

File: mono/metadata/jit-info.c [hunk 1]
```diff
@@ -118,7 +121,10 @@ jit_info_table_free (MonoJitInfoTable *table, gboolean duplicate)
 		MonoJitInfoTableChunk *chunk = table->chunks [i];
 		MonoJitInfo *tombstone;
 
-		if (--chunk->refcount > 0)
+		mono_atomic_dec_i32 (&chunk->refcount);
+		mono_memory_barrier ();
+
+		if (mono_atomic_load_i32 (&chunk->refcount) > 0)
```
Comment (lambdageek) [inline mono/metadata/jit-info.c]: Can't two threads get here and both see a value of 0?

**Comment by alexanderkyte** [comment]
If you crash the runtime with ```free(-1)```, this appears after about 60 times on average:

```
(lldb) process attach --pid 97419
btProcess 97419 stopped
* thread #1, name = 'tid_307', queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x7f8d5524eaa8)
    frame #0: 0x00000001099e6f47 mono-sgen`mono_get_hazardous_pointer(pp=0x00007f8d5524eaa8, hp=0x0000000109d57600, hazard_index=1) at hazard-pointer.c:208
   205 	
   206 		for (;;) {
   207 			/* Get the pointer */
-> 208 			p = *pp;
   209 			/* If we don't have hazard pointers just return the
   210 			   pointer. */
   211 			if (!hp)
Target 0: (mono-sgen) stopped.

Executable module set to "/Users/akyte/mono_merp/mono/mini/mono-sgen".
Architecture set to: x86_64h-apple-macosx.
(lldb) bt
warning: could not execute support code to read Objective-C class data in the process. This may reduce the quality of type information available.
* thread #1, name = 'tid_307', queue = 'com.apple.main-thread', stop reason = EXC_BAD_ACCESS (code=1, address=0x7f8d5524eaa8)
  * frame #0: 0x00000001099e6f47 mono-sgen`mono_get_hazardous_pointer(pp=0x00007f8d5524eaa8, hp=0x0000000109d57600, hazard_index=1) at hazard-pointer.c:208
    frame #1: 0x00000001097ebb2e mono-sgen`jit_info_table_chunk_index(chunk=0x00007f8c13e3a950, hp=0x0000000109d57600, addr=0x0000000109dcf2f8) at jit-info.c:200
    frame #2: 0x00000001097ea79a mono-sgen`jit_info_table_find(table=0x00007f8c13e3a520, hp=0x0000000109d57600, addr=0x0000000109dcf2f8) at jit-info.c:222
    frame #3: 0x00000001097ea586 mono-sgen`mono_jit_info_table_find_internal(domain=0x00007f8c13e0d460, addr=0x0000000109dcf2f8, try_aot=1, allow_trampolines=0) at jit-info.c:293
    frame #4: 0x00000001095963eb mono-sgen`mini_jit_info_table_find_ext(domain=0x00007f8c13e0d460, addr=0x0000000109dcf2f8, allow_trampolines=0, out_domain=0x0000000000000000) at mini-exceptions.c:1865
    frame #5: 0x0000000109598519 mono-sgen`mini_jit_info_table_find(domain=0x00007f8c13e0d460, addr=0x0000000109dcf2f8, out_domain=0x0000000000000000) at mini-exceptions.c:1903
    frame #6: 0x0000000109695463 mono-sgen`mono_arch_unwind_frame(domain=0x00007f8c13e0d460, jit_tls=0x00007f8c14841200, ji=0x0000000000000000, ctx=0x00007ffee677efa8, new_ctx=0x00007ffee677ee20, lmf=0x00007ffee677f138, save_locations=0x0000000000000000, frame=0x00007ffee677edb0) at exceptions-amd64.c:692
    frame #7: 0x00000001095967c2 mono-sgen`arch_unwind_frame(domain=0x00007f8c13e0d460, jit_tls=0x00007f8c14841200, ji=0x0000000000000000, ctx=0x00007ffee677efa8, new_ctx=0x00007ffee677ee20, lmf=0x00007ffee677f138, save_locations=0x0000000000000000, frame=0x00007ffee677edb0) at mini-exceptions.c:460
    frame #8: 0x0000000109596081 mono-sgen`mono_find_jit_info_ext(domain=0x00007f8c13e0d460, jit_tls=0x00007f8c14841200, prev_ji=0x0000000000000000, ctx=0x00007ffee677efa8, new_ctx=0x00007ffee677ee20, trace=0x0000000000000000, lmf=0x00007ffee677f138, save_locations=0x0000000000000000, frame=0x00007ffee677edb0) at mini-exceptions.c:648
    frame #9: 0x0000000109598737 mono-sgen`unwinder_unwind_frame(unwinder=0x00007ffee677ed50, domain=0x00007f8c13e0d460, jit_tls=0x00007f8c14841200, prev_ji=0x0000000000000000, ctx=0x00007ffee677efa8, new_ctx=0x00007ffee677ee20, trace=0x0000000000000000, lmf=0x00007ffee677f138, save_locations=0x0000000000000000, frame=0x00007ffee677edb0) at mini-exceptions.c:769
    frame #10: 0x000000010959792e mono-sgen`mono_walk_stack_full(func=(mono-sgen`summarize_frame at mini-exceptions.c:1575), start_ctx=0x00007ffee677f930, domain=0x00007f8c13e0d460, jit_tls=0x00007f8c14841200, lmf=0x00007ffee6780c98, unwind_options=MONO_UNWIND_LOOKUP_IL_OFFSET, user_data=0x00007ffee677f2a8, crash_context=1) at mini-exceptions.c:1266
    frame #11: 0x0000000109594e62 mono-sgen`mono_summarize_managed_stack(out=0x0000000109fed000) at mini-exceptions.c:1634
    frame #12: 0x0000000109889147 mono-sgen`summarizer_state_term(state=0x0000000109ae3070, out=0x00007ffee677faf0, mem="", provided_size=500000, controlling=0x0000000109fed000) at threads.c:6385
    frame #13: 0x0000000109888a44 mono-sgen`mono_threads_summarize_execute(ctx=0x00007ffee677f930, out=0x00007ffee677faf0, hashes=0x00007ffee677fae0, silent=0, working_mem="", provided_size=500000) at threads.c:6472
    frame #14: 0x0000000109889415 mono-sgen`mono_threads_summarize(ctx=0x00007ffee677f930, out=0x00007ffee677faf0, hashes=0x00007ffee677fae0, silent=0, signal_handler_controller=1, mem="", provided_size=500000) at threads.c:6527
    frame #15: 0x00000001096add17 mono-sgen`dump_native_stacktrace(signal="SIGABRT", ctx=0x00007ffee6780ab0) at mini-posix.c:1036
    frame #16: 0x00000001096ad8df mono-sgen`mono_dump_native_crash_info(signal="SIGABRT", ctx=0x00007ffee6780ab0, info=0x00007ffee6780a48) at mini-posix.c:1159
    frame #17: 0x000000010959b615 mono-sgen`mono_handle_native_crash(signal="SIGABRT", ctx=0x00007ffee6780ab0, info=0x00007ffee6780a48) at mini-exceptions.c:3265
    frame #18: 0x00000001096accb0 mono-sgen`sigabrt_signal_handler(_dummy=6, _info=0x00007ffee6780a48, context=0x00007ffee6780ab0) at mini-posix.c:232
    frame #19: 0x00007fff7dc55f5a libsystem_platform.dylib`_sigtramp + 26
    frame #20: 0x00007fff7da97b67 libsystem_kernel.dylib`__pthread_kill + 11
    frame #21: 0x00007fff7dc62080 libsystem_pthread.dylib`pthread_kill + 333
    frame #22: 0x00007fff7d9f31ae libsystem_c.dylib`abort + 127
    frame #23: 0x00007fff7daf1822 libsystem_malloc.dylib`free + 521
    frame #24: 0x0000000109f2d940 libtest.0.dylib`monoeg_g_free(ptr=0x00007ffee6780c54) at gmem.c:83
    frame #25: 0x0000000109f2739b libtest.0.dylib`mono_test_MerpCrashMalloc at libtest.c:7711
    frame #26: 0x0000000109dcf2f8
    frame #27: 0x0000000109dcb071
    frame #28: 0x00000001094965b3 mono-sgen`mono_jit_runtime_invoke(method=0x00007f8c13e0ebc8, obj=0x0000000000000000, params=0x00007ffee67811a0, exc=0x00007ffee6780f28, error=0x00007ffee6781270) at mini-runtime.c:3225
    frame #29: 0x000000010984c76d mono-sgen`do_runtime_invoke(method=0x00007f8c13e0ebc8, obj=0x0000000000000000, params=0x00007ffee67811a0, exc=0x0000000000000000, error=0x00007ffee6781270) at object.c:2993
    frame #30: 0x0000000109844f31 mono-sgen`mono_runtime_invoke_checked(method=0x00007f8c13e0ebc8, obj=0x0000000000000000, params=0x00007ffee67811a0, error=0x00007ffee6781270) at object.c:3161
    frame #31: 0x00000001098521b8 mono-sgen`do_exec_main_checked(method=0x00007f8c13e0ebc8, args=0x000000010a0003e8, error=0x00007ffee6781270) at object.c:5063
    frame #32: 0x000000010984ff03 mono-sgen`mono_runtime_exec_main_checked(method=0x00007f8c13e0ebc8, args=0x000000010a0003e8, error=0x00007ffee6781270) at object.c:5159
    frame #33: 0x000000010984ff56 mono-sgen`mono_runtime_run_main_checked(method=0x00007f8c13e0ebc8, argc=2, argv=0x00007ffee6781778, error=0x00007ffee6781270) at object.c:4616
    frame #34: 0x000000010954637f mono-sgen`mono_jit_exec_internal(domain=0x00007f8c13e0d460, assembly=0x00007f8c13e22fa0, argc=2, argv=0x00007ffee6781778) at driver.c:1298
    frame #35: 0x00000001095461ad mono-sgen`mono_jit_exec(domain=0x00007f8c13e0d460, assembly=0x00007f8c13e22fa0, argc=2, argv=0x00007ffee6781778) at driver.c:1257
    frame #36: 0x000000010954adef mono-sgen`main_thread_handler(user_data=0x00007ffee67816c0) at driver.c:1375
    frame #37: 0x00000001095490b2 mono-sgen`mono_main(argc=3, argv=0x00007ffee6781770) at driver.c:2580
    frame #38: 0x00000001094800ae mono-sgen`mono_main_with_options(argc=3, argv=0x00007ffee6781770) at main.c:50
    frame #39: 0x000000010947f6bd mono-sgen`main(argc=3, argv=0x00007ffee6781770) at main.c:408
    frame #40: 0x00007fff7d947015 libdyld.dylib`start + 1
(lldb) up
frame #1: 0x00000001097ebb2e mono-sgen`jit_info_table_chunk_index(chunk=0x00007f8c13e3a950, hp=0x0000000109d57600, addr=0x0000000109dcf2f8) at jit-info.c:200
   197 	
   198 		while (left < right) {
   199 			int pos = (left + right) / 2;
-> 200 			MonoJitInfo *ji = (MonoJitInfo *)mono_get_hazardous_pointer((gpointer volatile*)&chunk->data [pos], hp, JIT_INFO_HAZARD_INDEX);
   201 			gint8 *code_end = (gint8*)ji->code_start + ji->code_size;
   202 	
   203 			if (addr < code_end)
(lldb) p pos
(int) $0 = 673720360
(lldb) list 180
   180 	
   181 			if (addr < chunk->last_code_end)
   182 				right = pos;
   183 			else
   184 				left = pos + 1;
   185 		} while (left < right);
   186 		g_assert (left == right);
   187 	
   188 		if (left >= table->num_chunks)
   189 			return table->num_chunks - 1;
(lldb) list
   190 		return left;
   191 	}
   192 	
   193 	static int
   194 	jit_info_table_chunk_index (MonoJitInfoTableChunk *chunk, MonoThreadHazardPointers *hp, gint8 *addr)
   195 	{
   196 		int left = 0, right = chunk->num_elements;
   197 	
   198 		while (left < right) {
   199 			int pos = (left + right) / 2;
(lldb) list
   200 			MonoJitInfo *ji = (MonoJitInfo *)mono_get_hazardous_pointer((gpointer volatile*)&chunk->data [pos], hp, JIT_INFO_HAZARD_INDEX);
   201 			gint8 *code_end = (gint8*)ji->code_start + ji->code_size;
   202 	
   203 			if (addr < code_end)
   204 				right = pos;
   205 			else
   206 				left = pos + 1;
   207 		}
   208 		g_assert (left == right);
   209 	
(lldb) p chunk->num_elements
(volatile int) $1 = 1347440720
(lldb) up
frame #2: 0x00000001097ea79a mono-sgen`jit_info_table_find(table=0x00007f8c13e3a520, hp=0x0000000109d57600, addr=0x0000000109dcf2f8) at jit-info.c:222
   219 		chunk_pos = jit_info_table_index (table, (gint8*)addr);
   220 		g_assert (chunk_pos < table->num_chunks);
   221 	
-> 222 		pos = jit_info_table_chunk_index (table->chunks [chunk_pos], hp, (gint8*)addr);
   223 	
   224 		/* We now have a position that's very close to that of the
   225 		   first element whose end address is higher than the one
(lldb) p table->chunks [0]->num_elements
(volatile int) $2 = 1347440720
(lldb) p table->chunks [1]->num_elements
(volatile int) $3 = 56
(lldb) p table->chunks [2]->num_elements
(volatile int) $4 = 39
(lldb) p table->chunks [3]->num_elements
error: Couldn't apply expression side effects : Couldn't dematerialize a result variable: couldn't read its memory
(lldb) p table->num_chunks
(int) $6 = 3
```

**Comment by alexanderkyte** [comment]
I am closing this PR because the root cause of the crashes seen here was diagnosed to be internal to the system malloc. The 0x50505050505 pattern is seen across a number of unrelated examples of memory-unsafety with this test. 

The lack of watchpoint triggering is explained by the memory corruption happening through a different virtual address, as an implementation detail of the allocator. It can be seen by allocating a lot of zeroed memory in the telemetry enablement function, and then probing it after the crash to see the 0x505050505050 repeating pattern.

That is one of the reproductions seen before, with the strdup-allocated MERP strings becoming "PPPPPPP", which is the char representation of that bit pattern. 

https://github.com/mono/mono/pull/12864

**Pull request closed without merging**